### PR TITLE
You can milk snakes (and refactored milking to a component)

### LINKED ISSUE #7434: Make it possible to milk snakes!
# Milking Snakes
We have long been able to milk cows for milk.
I propose another feature: Milking snakes for venom!
Simply use a chem container like a shot glass on a snake to get toxin similar to a cow!

This is a real thing:

https://youtu.be/7ziWrneMYss?t=142

# Why its good for the game
Its a very small thing with little abuse potential (since you have to go through the effort of getting a snake either from maint or cargo first only to get a bit of toxin, if you wanted to use it maliciously it would be far easier to just buy some from med.
However, it would make for a very cool oozeling/ slimeperson gimmick! Having your own source of toxins to replenish some slime blood at any time is pretty cool.

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
Snakes can now be milked! AND, you can make anything milkable with a simple component applied!

## Why It's Good For The Game
Closes #7434, its pretty cool honestly, helps any future coders immensely if they want to make their new mob milkable.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

(milked multiple time bc i messed up chat)
image

</details>

## Changelog
:cl:
add: Snakes can be milked for toxin
refactor: Milkability is a component now
/:cl:

### CODE REVIEW

**Comment by Tyranicranger4** [comment]
I'm going to find a way to VV that onto a moth. I don't think Beestation is ready for this future.

**Comment by Rukofamicom** [comment]
Dexity awaits the day @Tyranicranger4

File: code/datums/components/milkable.dm [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	if(transfered)
+		user.visible_message("[user] milks [parent] using \the [container].", "<span class='notice'>You milk [parent] using \the [container].</span>")
+	else
+		to_chat(user, "<span class='danger'>The udder is dry. Wait a bit longer...</span>")
+
+	if(on_milked)
+		on_milked.Invoke()
+
+/datum/component/milkable/proc/restart_gen()
+	START_PROCESSING(SSdcs, src)
```
Comment (Archanial) [inline code/datums/components/milkable.dm:69]: Missing signal handler

**Review by Archanial — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
I don't know component well, but I think this doesn't qdel its created component when it is deleted.

File: code/datums/components/milkable.dm [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// how cows and other simplemobs are able to be milked
+
+/datum/component/milkable
+	var/mob/producer
+	var/obj/item/reagent_containers/intstorage
+	var/max_reagents
+	var/min_reagents
+	var/datum/reagent/reagent_generated
+	var/datum/callback/on_milked
```
Comment (Archanial) [inline code/datums/components/milkable.dm]: This callback is never used
If the use of it is ever needed you can just readd it

Comment (Archanial) [reply code/datums/components/milkable.dm]: Same goes to containers

**Review by Archanial — CHANGES_REQUESTED** [review]

File: code/datums/components/milkable.dm [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	var/list/allowed_containers = list(/obj/item/reagent_containers/glass)
+
+/datum/component/milkable/Initialize(_max_reagents, _starting_amount, _reagent_used, _callback_milked, _allowed_containers)
+	if(!ismob(parent))
```
Comment (Archanial) [inline code/datums/components/milkable.dm:13]: Add ishuman check to this

**Review by Archanial — CHANGES_REQUESTED** [review]

**Review by PowerfulBacon — COMMENTED** [review]
I think this component needs to be more general:
 - Snakes shouldn't have udder messages.
 - Snakes should have a different extraction method, direct syringe extraction rather than just using buckets.

Some other comments:
 - Process can be removed entirely, you can determine functionally how much reagent something should have based solely on the time passed since last extraction and how much was left last extraction
 - Extraction should only work on help intent. Try and attack a cow with a poison knife for example, and you will get filled with milk.
 - There should be a do_after timer somewhere I think, you'll have to figure out how to manage that without sleeping directly in the extraction proc.

**Comment by EvilDragonfiend** [comment]
> Snakes should have a different extraction method, direct syringe extraction rather than just using buckets.

I'd prefer to see `touch a snake into a beaker` than `touch a beaker to a snake`, because that's how you extract venom from a snake in the real world (although you need to cover plastic wrap on a beaker).

File: code/datums/components/milkable.dm (hunk 3 above)
Comment (etherware-novice) [reply code/datums/components/milkable.dm:13]: arent humans mobs?

**Comment by etherware-novice** [comment]
>  - Process can be removed entirely, you can determine functionally how much reagent something should have based solely on the time passed since last extraction and how much was left last extraction

about this, it might make the code inconsistant with current mobs generating it on life. its also really really confusing to simulate 10 prob(90) calls in the milking proc, and youd also have to take into account time spent dead..

**Comment by etherware-novice** [comment]
i think a good compromise might be sending a COMSIG_MOB_LIFE signal whenever life() is called, as that would also simplify checking for deathstates

File: code/datums/components/milkable.dm (hunk 3 above)
Comment (Archanial) [reply code/datums/components/milkable.dm:13]: This is the point
I don't want to see milkable on humans

Comment (etherware-novice) [reply code/datums/components/milkable.dm:13]: idk..if a senior admin wants to do something with your code it would be rude to make it impossible right

Comment (etherware-novice) [reply code/datums/components/milkable.dm:13]: ~~also i want to see the chaos that ensues~~

**Comment by FriendlyContractor** [comment]
While there sadly is no official design documentation i believe everyone can agree that reducing potential for griefing and powergaming are our objectives. This PR contradicts them both by enabling players to obtain dangerous chemical they have no valid need for and thus tempts them to act against the rules and spirit of the server. Conclusion is that milking snakes for venom should either be **MECHANICALY** restricted to antagonists who might have a valid use for it or simply or not added at all as its going to be a source of problems (_botany *wink *wink_)

**Comment by FriendlyContractor** [comment]
I think you completly missed the point.

**Comment by etherware-novice** [comment]
uhh fine then traitors have training to do it properly and standard crew just get nothing

**Comment by FriendlyContractor** [comment]
> uhh fine then traitors have training to do it properly and standard crew just get nothing

Does not have to only be traitors. Anyone from lowly revolutionary to wizard is fine and im pretty sure there is even a convinient function to check if person is an antagonist. Also make sure failure message for normal crew members hints at a fact it could be done, were they up to no good.

**Comment by etherware-novice** [comment]
okey

also revs are NOT lowly

**Comment by itsmeow** [comment]
I'm not really concerned with this PR and its potential for griefing. Not everyone has access to snakes, and if someone chooses to get snake venom, how they use it is up to their character. If they grief, they suffer the consequences. Getting deadly chemicals as a non-antag is not supposed to be *impossible*, just discouraged.

**Comment by St0rmC4st3r** [comment]
I love it. Time to reimplement the Enlarged Giant Spider into an actual separate mob, rather than a map edit.

**Comment by etherware-novice** [comment]
giant enemy spider (also wow coincidence the other pr i did yeterday was for subclassing a map vv thing)

File: code/datums/components/milkable.dm (hunk 3 above)
Comment (dwasint) [reply code/datums/components/milkable.dm:13]: thats not chaos thats just erp i agree ishuman checks to stop humans from being milked

File: code/datums/components/milkable.dm [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		COMSIG_PARENT_ATTACKBY,
+		COMSIG_LIVING_REVIVE))
+	STOP_PROCESSING(SSdcs, src)
+
+
+/datum/component/milkable/process(delta_time)
+	if(producer.stat == DEAD)
+		return PROCESS_KILL
+
+	if(producer.stat != CONSCIOUS || prob(90))
```
Comment (dwasint) [inline code/datums/components/milkable.dm:41]: Why do things need to be CONSCIOUS to produce milk, i don't think its an action that is consciously thought seems like a weird restriction

File: code/datums/components/milkable.dm [hunk 5]
```diff
@@ -0,0 +1,67 @@
+// how cows and other simplemobs are able to be milked
+
+/datum/component/milkable
+	var/mob/producer
+	var/obj/item/reagent_containers/intstorage
```
Comment (dwasint) [inline code/datums/components/milkable.dm]: also please seperate these nitpicking i know but initial_storage or int_storage looks much more pleasing to look at than intstorage

File: code/datums/components/milkable.dm (hunk 4 above)
Comment (etherware-novice) [reply code/datums/components/milkable.dm:41]: idk but the original code for it had that check so i implemented it here

**Comment by ImSynthex** [comment]
can the snakes milk you back? (asking for scientific reasons)

**Comment by ivanmixo** [comment]
_Please no._

**Comment by BrumalLake** [comment]
really thought it was a port of this https://github.com/tgstation/tgstation/pull/58910

File: code/datums/components/milkable.dm (hunk 3 above)
Comment (etherware-novice) [reply code/datums/components/milkable.dm:13]: i added like 3 different "yes im sure" checks to the addcomponent milkable, should be good now

**Comment by Rukofamicom** [comment]
image

What is happening, I don't like this timeline.

**Pull request closed without merging**